wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceRuntimes.java

```diff
@@ -245,6 +246,74 @@ public void stop(String workspaceId) throws NotFoundException, ServerException,
     }
 
     /**
+     * Adds machine into running workspace.
+     * This method do not touch workspace configuration.
+     * Just adds machine to workspace runtime and destroy it on workspace stop.
+     *
+     * @param machine
+     *         machine to add to specified runtime
+     * @throws NotFoundException
+     *         when workspace with specified id not running or not exists
+     * @throws ServerException
+     *         when application server is stopping
+     * @throws ConflictException
+     *         when workspace is not running
+     */
+    public void addMachine(MachineImpl machine) throws NotFoundException, ServerException, ConflictException {
+        ensurePreDestroyIsNotExecuted();
+
+        String workspaceId = machine.getWorkspaceId();
+
+        rwLock.writeLock().lock();
+        try {
+            ensurePreDestroyIsNotExecuted();
+            final RuntimeDescriptor descriptor = descriptors.get(workspaceId);
+            if (descriptor == null) {
+                throw new NotFoundException("Workspace with id '" + workspaceId + "' is not running.");
+            }
+            if (descriptor.getRuntimeStatus() != WorkspaceStatus.RUNNING) {
+                throw new ConflictException("Cannot add machine " + machine.getId() + " to not running workspace.");
+            }
+
+            descriptor.getRuntime().getMachines().add(machine);
+        } finally {
+            rwLock.writeLock().unlock();
+        }
+    }
+
+    /**
+     * Removes machine from running workspace.
+     * This method do not touch workspace configuration and is opposite to {@link #addMachine(MachineImpl)}
+     * Does nothing if remove nonexistent machine
+     *
+     * @param machine
+     *         machine to remove from specified runtime
+     * @throws NotFoundException
+     *         when workspace with specified id not running or not exists
+     */
+    public void removeMachine(MachineImpl machine) throws NotFoundException {
+        String workspaceId = machine.getWorkspaceId();
+
+        rwLock.writeLock().lock();
+        try {
+            final RuntimeDescriptor descriptor = descriptors.get(workspaceId);
+            if (descriptor == null) {
+                throw new NotFoundException("Workspace with id '" + workspaceId + "' is not running.");
+            }
+
+            List<MachineImpl> machines = descriptor.getRuntime().getMachines();
+            for (int i = 0; i < machines.size(); i++) {
+                if (machine.getId().equals(machines.get(i).getId())) {
+                    machines.remove(i);
```

Comment:
No need in such detailed explanation. Your point is almost correct! However next time please use common terminology to avoid misunderstanding: [Big O notation](https://en.wikipedia.org/wiki/Big_O_notation), [Worst-case complexity](https://en.wikipedia.org/wiki/Worst-case_complexity), etc.